test/Validate.CFG.cpp

```diff
@@ -0,0 +1,601 @@
+               " %func    = OpFunction %voidt None %funct\n";
+  str += entry >> cont;
+  str += cont >> exit;
+  str += exit;
+  str += "OpFunctionEnd\n";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_SUCCESS, ValidateInstructions());
+}
+
+TEST_F(ValidateCFG, VariableNotInFirstBlockBad) {
+  Block entry("entry");
+  Block cont("cont");
+  Block exit("exit", SpvOpReturn);
+
+  // This operation should only be performed in the entry block
+  cont.setBody("%var = OpVariable %ptrt Function\n");
+
+  string str = header + nameOps(make_pair("func", "Main")) + types_consts +
+               " %func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> cont;
+  str += cont >> exit;
+  str += exit;
+  str += "OpFunctionEnd\n";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_CFG, ValidateInstructions());
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("first block"));
+}
+
+TEST_F(ValidateCFG, BlockAppearsBeforeDominatorBad) {
+  Block entry("entry");
+  Block cont("cont");
+  Block branch("branch", SpvOpBranchConditional);
+  Block merge("merge", SpvOpReturn);
+
+  branch.setBody(
+      " %cond    = OpSLessThan %intt %one %two\n"
+      "OpSelectionMerge %merge None\n");
+
+  string str = header + nameOps("cont", "branch", make_pair("func", "Main")) +
+               types_consts + "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> branch;
+  str += cont >> merge;  // cont appears before its dominator
+  str += branch >> vector<Block>({cont, merge});
+  str += merge;
+  str += "OpFunctionEnd\n";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_CFG, ValidateInstructions());
+  EXPECT_THAT(getDiagnosticString(),
+              HasSubstr("appears in the binary before its dominator"));
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("cont"));
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("branch"));
+}
+
+TEST_F(ValidateCFG, MergeBlockTargetedByMultipleHeaderBlocksBad) {
+  Block entry("entry");
+  Block loop("loop");
+  Block badhead("badhead", SpvOpBranchConditional);
+  Block t("t");
+  Block f("f");
+  Block cont("cont");
+  Block merge("merge");
+  Block end("end", SpvOpReturn);
+
+  // cannot share the same merge
+  loop.setBody(
+      " %cond2   = OpSLessThan %intt %one %two\n"
+      " OpLoopMerge %merge %cont None\n");
+  badhead.setBody(
+      " %cond1   = OpSLessThan %intt %one %two\n"
+      "OpSelectionMerge %merge None\n");
+
+  string str = header + nameOps("merge", make_pair("func", "Main")) + types_consts +
+               "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> loop;
+  str += loop >> badhead;
+  str += badhead >> vector<Block>({t, f});
+  str += t >> merge;
+  str += f >> cont;
+  str += cont >> loop;
+  str += merge >> end;
+  str += end;
+  str += "OpFunctionEnd\n";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_CFG, ValidateInstructions());
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("merge"));
```

Comment:
This substring is too vague.  Isn't there a way to assert with higher certainty that the validation fails specifically because of `%merge` being reused by multiple header blocks?